### PR TITLE
Add read abstraction to the State

### PR DESCRIPTION
**Pull request by DarianShawn**
# Description

It adds a new abstraction for the State such that instead of having access to a low-level interface with a method Get to query for account and storage, now, we get a much richer abstraction with two methods:

```
GetAccount(...) (*Account, error)
GetStorage(...) (Hash, error)
```

There are several benefits of this new data layer. 
First, the layers that want to access the state do not have to perform extra computation on top of the raw result of Get to parse the objects.
Second, it gives more flexibility to the state storage to decide how it wants to store information and not force it to adapt to the Get interface.
Third, we do return error and handle it correctly instead of printing it in console only. 

The PR is heavily expiring by Polygon Edge [PR 827](https://github.com/0xPolygon/polygon-edge/pull/827)

# Changes include

- [x] Bugfix (non-breaking change that solves an issue)
- [x] New feature (non-breaking change that adds functionality)

## Testing

- [x] I have tested this code with the official test suite

### CODE REVIEW

File: state/immutable-trie/snapshot.go [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+	val, err := snapshot.trie.Get(key, s.state)
+	if err != nil {
+		// something bad happen, should not continue
+		return types.Hash{}, err
+	} else if len(val) == 0 {
+		// not found
+		return types.Hash{}, nil
+	}
+
+	p := &fastrlp.Parser{}
```
Comment (0xcb9ff9) [inline state/immutable-trie/snapshot.go:51]: use `fastrlp.DefaultParserPool`  reuse object ?

Comment (DarianShawn) [reply state/immutable-trie/snapshot.go:51]: The variable is on stack, will not escape to heap.

**Review by 0xcb9ff9 — DISMISSED** [review]
LGMT

File: state/immutable-trie/snapshot.go (hunk 1 above)
Comment (0xcb9ff9) [reply state/immutable-trie/snapshot.go:51]: ok.

**Review by abrahamcruise321 — DISMISSED** [review]
LGTM.

**Review by abrahamcruise321 — APPROVED** [review]
LGTM.

**Pull request merged**